tempesta_fw/tls.c

```diff
@@ -569,6 +570,43 @@ tfw_tls_conn_drop(TfwConn *c)
 	tfw_conn_hook_call(TFW_FSM_HTTP, c, conn_drop);
 }
 
+int tfw_tls_skb_copy_pages(struct sk_buff *skb)
+{
+	int i;
+	struct sk_buff *frag_iter;
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		skb_frag_t *f = &skb_shinfo(skb)->frags[i];
+		struct page *p, *p2;
+		int order;
+		u8 *vaddr, *vaddr2;
+		unsigned int frag_size = skb_frag_size(f);
+
+		p = skb_frag_page(f);
+		order = get_order(frag_size);
+		p2 = alloc_pages(__GFP_MEMALLOC, order);
+		if (!p2)
+			goto err;
+		vaddr = kmap_atomic(p);
+		vaddr2 = kmap_atomic(p2);
+		memcpy(vaddr2, vaddr + f->page_offset, frag_size);
```

Comment:
Currently, encryption overwrites data in-place. But we may create another scatter-gather list for destination, and avoid copying data here. I'm talking about `aead_request_set_crypt(req, sgt->sgl, sgt->sgl, elen, xfrm->iv_enc);` in `ttls_encrypt()`. As you can see, currently third argument which is destination, is the same as second (source).

Reply:
Yes, this is definitely the better approach. If TCP/IP stack clears `SKBTX_SHARED_FRAG`, then we can use `tempesta_tls_skb_settype()` to pass the knowledge about the skb nature through the stack. Also probably `ttls_encrypt()` should accept an argument controlling how to perform the encryption - with or without copying.